src/dns.rs

```diff
@@ -0,0 +1,305 @@
+use core::mem;
+
+use anyhow::*;
+use log::*;
+
+use esp_idf_sys::c_types::*;
+use esp_idf_sys::*;
+
+use crate::private::cstr::{CStr, CString};
+use crate::task::{TaskConfig, TaskHandle};
+use crate::lwip;
+
+/// 0.0.0.0
+const IPADDR_ANY: u32 = 0x00000000;
+
+const DNS_PORT: c_ushort = 53;
+const DNS_MAX_LEN: usize = 256;
+
+const OPCODE_MASK: u16 = 0x7800;
+const QR_FLAG: u16 = 1 << 7;
+const QD_TYPE_A: u16 = 0x0001;
+const ANS_TTL_SEC: u32 = 300;
+
+pub struct CaptivePortalDns {
+    task_handle: Option<TaskHandle>,
+}
+
+impl CaptivePortalDns {
+    pub fn new() -> Self {
+        CaptivePortalDns { task_handle: None }
+    }
+
+    pub fn start(&mut self) -> Result<()> {
+        if self.task_handle.is_some() {
+            bail!("dns server is already running");
+        }
+
+        let handle = TaskConfig::default()
+            .priority(5)
+            .spawn("dns_server", dns_server_task)?;
+
+        self.task_handle = Some(handle);
+        Ok(())
+    }
+
+    pub fn stop(&mut self) -> Result<()> {
+        if let Some(handle) = self.task_handle.take() {
+            handle.stop();
+            Ok(())
+        } else {
+            Err(anyhow!("dns task already stopped or was never started"))
+        }
+    }
+}
+
+impl Drop for CaptivePortalDns {
+    fn drop(&mut self) {
+        if self.task_handle.is_some() {
+            self.stop().unwrap();
+        }
+    }
+}
+
+/// DNS Header Packet
+#[repr(C, packed)]
+struct DnsHeader {
+    id: u16,
+    flags: u16,
+    qd_count: u16,
+    an_count: u16,
+    ns_count: u16,
+    ar_count: u16,
+}
+
+/// DNS Question Packet
+#[repr(C)]
+struct DnsQuestion {
+    typ: u16,
+    class: u16,
+}
+
+/// DNS Answer Packet
+#[repr(C, packed)]
+struct DnsAnswer {
+    name_ptr: u16,
+    typ: u16,
+    class: u16,
+    ttl: u32,
+    addr_len: u16,
+    ip_addr: u32,
+}
+
+fn parse_dns_name(raw_name: *mut u8, parsed_name: &mut [u8]) -> *mut u8 {
```

Comment:
This and the next function are littered with `unsafe` statements. Would it be possible to stay in Rust safe-land for parsing this stuff? Only if possible?